cargo-miri/bin.rs

```diff
@@ -251,26 +251,72 @@ fn xargo_check() -> Command {
 
 /// Execute the command. If it fails, fail this process with the same exit code.
 /// Otherwise, continue.
-fn exec(mut cmd: Command) {
-    let exit_status = cmd.status().expect("failed to run command");
-    if exit_status.success().not() {
-        std::process::exit(exit_status.code().unwrap_or(-1))
+fn exec(mut cmd: Command) -> ! {
+    // On non-Unix imitate POSIX exec as closely as we can
+    #[cfg(not(unix))]
+    {
+        let exit_status = cmd.status().expect("failed to run command");
+        std::process::exit(exit_status.code().unwrap_or(-1));
+    }
+    // On Unix targets, actually exec
+    // If exec returns, process setup has failed. This is the same error condition as the expect in
+    // the non-Unix case.
+    #[cfg(unix)]
+    {
+        use std::os::unix::process::CommandExt;
+        let error = cmd.exec();
+        Err(error).expect("failed to run command")
     }
 }
 
 /// Execute the command and pipe `input` into its stdin.
 /// If it fails, fail this process with the same exit code.
 /// Otherwise, continue.
 fn exec_with_pipe(mut cmd: Command, input: &[u8]) {
-    cmd.stdin(process::Stdio::piped());
-    let mut child = cmd.spawn().expect("failed to spawn process");
+    #[cfg(unix)]
     {
-        let stdin = child.stdin.as_mut().expect("failed to open stdin");
-        stdin.write_all(input).expect("failed to write out test source");
+        use std::os::unix::io::FromRawFd;
+        use std::process::Stdio;
+
+        let mut fds: [libc::c_int; 2] = [0, 0];
+        // SAFETY: fds is an array of 2 libc::c_int
+        let res = unsafe { libc::pipe(fds.as_mut_ptr()) };
+        assert_eq!(res, 0, "failed to create pipe");
+
+        // We need to set close-on-exec, otherwise our pipe isn't readable after exec
```

Comment:
This comment makes no sense to me... close-on-exec would make it *less* readable after exec, by closing it?

Reply:
I'm implementing the same logic as here: https://github.com/rust-lang/miri/blob/a591e9fc5b6c382ced52c217e02e595e0b598c65/cargo-miri/bin.rs#L267-L270

We need to close the file descriptor we've opened for writing so that the child can read from the pipe.

For what it's worth, I've worked that out by experimentation. The man page says this:
> If all file descriptors referring to the write end of a pipe have
been closed, then an attempt to [read(2)](https://man7.org/linux/man-pages/man2/read.2.html) from the pipe will see
end-of-file ([read(2)](https://man7.org/linux/man-pages/man2/read.2.html) will return 0).

which seems to contradict reality. But I know for sure that if this file descriptor isn't closed, either by a `close` or by setting close-on-exec (which will make the exec close it), the child process just blocks forever trying to read from its own file descriptor 0.

Reply:
> We need to close the file descriptor we've opened for writing so that the child can read from the pipe.

No that's not what this means. We can read from the pipe just fine without closing the FD.

However, if we don't close the FD, reading will never stop. It's like a crossbeam channel that is never closed. That's why you see the blocking. It reads all the input and *then* blocks because it has no idea that there is no more input to come.

Reply:
That makes a lot more sense. It still rather hurts my brain that `close` has semantics like this.